src/streaming.ts

```diff
@@ -46,45 +52,113 @@ export class StreamProcessor {
-      {
-        headers: {
-          Authorization: `Bearer ${this.jwtToken}`,
-          'API-Key': this.apiKey,
-        },
+    this.log.info('Starting new StreamProcessor');
+
+    const url = `${this.options.baseUrl}/stream?cluster=${this.cluster}`;
+
+    const options = {
+      headers: {
+        'Cache-Control': 'no-cache',
+        Accept: 'text/event-stream',
+        Authorization: `Bearer ${this.jwtToken}`,
+        'API-Key': this.apiKey,
       },
-    );
+    };
 
-    eventSource.onopen = (event: MessageEvent) => {
-      this.log.debug('Stream connected', event);
+    const onConnected = () => {
+      this.log.info(`SSE stream connected OK: ${url}`);
+      this.retryAttempt = 0;
+      this.readyState = StreamProcessor.CONNECTED;
       this.eventBus.emit(StreamEvent.CONNECTED);
-      // In case we've had to retry, set it back to false
-      this.isRetrying = false
     };
 
-    eventSource.onretrying = (event: MessageEvent) => {
-      this.log.error('Stream is trying to reconnect', event);
-      // Only inform the client once that we're retrying
-      if (!this.isRetrying) {
-        this.isRetrying = true
-        this.eventBus.emit(StreamEvent.RETRYING);
-      }
-    };
+    const onFailed = (msg: string) => {
+      this.retryAttempt += 1;
 
-    eventSource.onerror = (event: MessageEvent) => {
-      this.log.error('Unrecoverable error with stream, not retrying to connect: ', event);
-      this.eventBus.emit(StreamEvent.ERROR, event);
+      const delayMs = this.getRandomRetryDelayMs();
+      this.log.warn(`SSE disconnected: ${msg} will retry in ${delayMs}ms`);
+      this.readyState = StreamProcessor.RETRYING;
+      this.eventBus.emit(StreamEvent.RETRYING);
+
+      setTimeout(() => {
+        this.log.info('SSE retrying to connect');
+        this.connect(url, options, onConnected, onFailed);
+      }, delayMs);
     };
 
-    eventSource.addEventListener('*', (event: MessageEvent) => {
-      const msg: StreamMsg = JSON.parse(event.data);
+    this.connect(url, options, onConnected, onFailed);
+
+    this.eventBus.emit(StreamEvent.READY);
+  }
+
+  private getRandomRetryDelayMs(): number {
+    const delayMs = this.retryDelayMs * this.retryAttempt;
+    return Math.min(delayMs, 60000);
+  }
 
-      this.log.debug('Received event from stream: ', msg);
+  private connect(
+    url: string,
+    options: RequestOptions,
+    onConnected: () => void,
+    onFailed: (msg: string) => void,
+  ): void {
+    if (this.readyState === StreamProcessor.CONNECTED) {
+      this.log.debug('SSE already connected, skip retry');
+      return;
+    }
+
+    const isSecure = url.startsWith('https:');
+    this.log.debug('SSE HTTP start request', url);
+
+    (isSecure ? https : http)
+      .request(url, options, (res) => {
+        this.log.debug('SSE got HTTP response code', res.statusCode);
+
+        if (res.statusCode !== 200) {
+          onFailed(`HTTP code ${res.statusCode}`);
+          return;
+        }
```

Comment:
This is a wee bit too simplistic since it will work fine with any 2xx or probably 3xx. I think what was there before was a check for any 4xx or 5xx response to trigger the error. It'll probably be fine for now 👍🏻

Reply:
i can add 4xx/5xx here, it'll be a simple one line change.

looks like the eventsource lib had this

`  function isRetryableError (error) {
    if (error) {
      if (error.status) {
        var status = error.status
        return status === 500 || status === 502 || status === 503 || status === 504
      }
    }
    // Always retry on IO Errors
    return true
  }`

Reply:
just seeing this now, i went for

`if (res.statusCode >= 400 && res.statusCode <= 599)`